utilities/include/vec_dh.h

```diff
@@ -13,178 +13,318 @@
+  }
+
+  VecDH<T>& operator=(VecDH<T>&& other) {
+    if (!other.cacheModified) {
+      if (other.impl_.size() > 0)
+        impl_ = std::move(other.impl_);
+      else
+        impl_.clear();
+      implModified = true;
+      cacheModified = false;
+    } else {
+      cache = std::move(other.cache);
+      cacheModified = true;
+      implModified = false;
+    }
+    return *this;
+  }
+
+  int size() const {
+    if (!cacheModified)
+      return impl_.size();
+    return cache.size();
+  }
 
   void resize(int newSize, T val = T()) {
     bool shrink = size() > 2 * newSize;
-    if (device_valid_) {
-      device_.resize(newSize, val);
-      if (shrink) device_.shrink_to_fit();
-    }
-    if (host_valid_) {
-      host_.resize(newSize, val);
-      if (shrink) host_.shrink_to_fit();
+    if (cacheModified) {
+      cache.resize(newSize, val);
+      if (shrink) cache.shrink_to_fit();
+    } else {
+      impl_.resize(newSize, val);
+      if (shrink) impl_.shrink_to_fit();
+      implModified = true;
     }
   }
 
   void swap(VecDH<T>& other) {
-    host_.swap(other.host_);
-    device_.swap(other.device_);
-    thrust::swap(host_valid_, other.host_valid_);
-    thrust::swap(device_valid_, other.device_valid_);
+    if (!cacheModified && !other.cacheModified) {
+      implModified = true;
+      other.implModified = true;
+      impl_.swap(other.impl_);
+    } else {
+      syncCache();
+      other.syncCache();
+      cacheModified = true;
+      implModified = false;
+      other.cacheModified = true;
+      other.implModified = false;
+      cache.swap(other.cache);
+    }
   }
 
-  using IterD = typename thrust::device_vector<T>::iterator;
-  using IterH = typename thrust::host_vector<T>::iterator;
-  using IterDc = typename thrust::device_vector<T>::const_iterator;
-  using IterHc = typename thrust::host_vector<T>::const_iterator;
+  using IterD = typename thrust::universal_vector<T>::iterator;
+  using IterH = typename thrust::universal_vector<T>::iterator;
+  using IterDc = typename thrust::universal_vector<T>::const_iterator;
+  using IterHc = typename thrust::universal_vector<T>::const_iterator;
 
   IterH begin() {
-    RefreshHost();
-    device_valid_ = false;
-    return host_.begin();
+    syncImpl();
+    implModified = true;
+    return impl_.begin();
   }
 
   IterH end() {
-    RefreshHost();
-    device_valid_ = false;
-    return host_.end();
+    syncImpl();
+    implModified = true;
+    return impl_.end();
   }
 
   IterHc cbegin() const {
-    RefreshHost();
-    return host_.cbegin();
+    syncImpl();
+    return impl_.cbegin();
   }
 
   IterHc cend() const {
-    RefreshHost();
-    return host_.cend();
+    syncImpl();
+    return impl_.cend();
   }
 
   IterHc begin() const { return cbegin(); }
   IterHc end() const { return cend(); }
 
   IterD beginD() {
-    RefreshDevice();
-    host_valid_ = false;
-    return device_.begin();
+    syncImpl();
+    implModified = true;
+    return impl_.begin();
```

Comment:
Since this is now identical to `begin()`, what do you say we just remove all the `D` and `H` specifics from this class?

Reply:
No, currently due to the workaround, they are not identical: we need to track when to synchronize the cache. The cache is needed for faster push_back. Othwise it will take forever to populate the vector.

Currently the synchronization mechanism is similar to the old VecDH: std::vector for host, and universal vector for both host and device. But the std::vector will only be created when we attempt to do push_back, and immediately synchronized to the universal_vector and cleared when we need to access it from the device.

Until the performance issue is fixed in thrust, we still need to keep track of host and device.

Reply:
oh, it seems that some of the functions that have h/d suffix can be merged together. Will do that tomorrow.